### PR TITLE
fix: better handling of npm resolution occurring on workers

### PR DESCRIPTION
**Pull request by dsherret**
Closes https://github.com/denoland/deno/issues/24063
Closes https://github.com/denoland/deno/pull/24084 (via changes to text_encoding.rs)
Closes https://github.com/denoland/deno/issues/21989
Closes https://github.com/denoland/deno/issues/24051 (via deno_graph update)

### CODE REVIEW

File: cli/npm/managed/resolution.rs [hunk 1]
```diff
@@ -223,34 +223,42 @@ impl NpmResolution {
...
-  /// called by deno_graph's NpmResolver or for resolving packages in
-  /// a package.json
-  pub fn resolve_pkg_req_as_pending_with_info(
-    &self,
-    pkg_req: &PackageReq,
-    package_info: &NpmPackageInfo,
-  ) -> Result<PackageNv, NpmPackageVersionResolutionError> {
-    debug_assert_eq!(pkg_req.name, package_info.name);
+    debug_assert_eq!(pkg_req.name, pkg_info.name);
+    let _permit = self.update_queue.acquire().await;
```
Comment (dsherret) [inline cli/npm/managed/resolution.rs:232]: Bug fix here. Previously two threads could be modifying the snapshot at the same time.

File: cli/npm/managed/resolution.rs [hunk 2]
```diff
@@ -223,34 +223,42 @@ impl NpmResolution {
...
       pkg_req,
-      package_info,
+      pkg_info,
     )?;
     Ok(nv)
   }
 
+  pub async fn resolve_pkg_reqs_as_pending_with_info(
+    &self,
+    reqs_with_pkg_infos: &[(&PackageReq, Arc<NpmPackageInfo>)],
```
Comment (dsherret) [inline cli/npm/managed/resolution.rs:245]: We now resolve these in a batch.

File: cli/tools/repl/session.rs [hunk 3]
```diff
@@ -639,10 +639,11 @@ impl ReplSession {
           source_map: deno_ast::SourceMapOption::None,
           source_map_file: None,
           inline_sources: false,
-          keep_comments: false,
+          remove_comments: false,
```
Comment (dsherret) [inline cli/tools/repl/session.rs:642]: Let's keep comments in the output here.

File: runtime/shared.rs [hunk 4]
```diff
@@ -111,9 +111,9 @@ pub fn maybe_transpile_source(
     )?
     .into_source();
 
-  let maybe_source_map: Option<SourceMapData> = transpiled_source
-    .source_map
-    .map(|sm| sm.into_bytes().into());
+  let maybe_source_map: Option<SourceMapData> =
+    transpiled_source.source_map.map(|sm| sm.into());
+  let source_text = String::from_utf8(transpiled_source.source)?;
```
Comment (dsherret) [inline runtime/shared.rs:116]: We could avoid this `String::from_utf8` by not requiring a FastString in deno_core.

File: cli/util/text_encoding.rs
Comment (dsherret) [inline cli/util/text_encoding.rs:1]: I updated both of these to handle the source map not being on the last line.

File: cli/emit.rs [hunk 5]
```diff
@@ -172,16 +172,43 @@ impl Emitter {
...
+      | MediaType::Dmts
+      | MediaType::Dcts
+      | MediaType::Json
+      | MediaType::Wasm
+      | MediaType::TsBuildInfo
+      | MediaType::SourceMap
+      | MediaType::Unknown => {
+        // clear this specifier from the parsed source cache as it's now out of date
+        self.parsed_source_cache.free(specifier);
+        Ok(source_code)
```
Comment (dsherret) [inline cli/emit.rs:209]: We were accidentally transpiling files that didn't need to be transpiled.

File: cli/emit.rs [hunk 6]
```diff
@@ -172,16 +172,43 @@ impl Emitter {
...
+      | MediaType::Cts
+      | MediaType::Jsx
+      | MediaType::Tsx => {
+        let source_arc: Arc<str> = source_code.into();
+        let parsed_source = self
+          .parsed_source_cache
+          .remove_or_parse_module(specifier, source_arc, media_type)?;
+        // HMR doesn't work with embedded source maps for some reason, so set
+        // the option to not use them (though you should test this out because
+        // this statement is probably wrong)
```
Comment (dsherret) [inline cli/emit.rs:187]: I think this might be due to accidentally transpiling JS

File: cli/util/text_encoding.rs
Comment (dsherret) [inline cli/util/text_encoding.rs:1]: These were both updated to handle the source map not being on the last line.

(I previously posted this comment, but it got outdated)

File: cli/npm/managed/cache/value_creator.rs [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+/// Attempts to create a shared value asynchronously on one tokio runtime while
+/// many runtimes are requesting the value.
+///
+/// This is only useful when the value needs to get created once across
+/// many runtimes.
+///
+/// This handles the case where one tokio runtime goes down while another
+/// one is still running.
+#[derive(Debug)]
+pub struct MultiRuntimeAsyncValueCreator<TResult: Send + Clone + 'static> {
```
Comment (dsherret) [inline cli/npm/managed/cache/value_creator.rs:33]: I resisted the urge to call this `MrValueCreator`.

File: cli/npm/managed/resolution.rs (hunk 1 above)
Comment (dsherret) [reply cli/npm/managed/resolution.rs:232]: I'm just creating an `SyncReadAsyncWriteLock` to make this mistake impossible in the future (will land in a future pr).

File: cli/emit.rs (hunk 5 above)
Comment (bartlomieju) [reply cli/emit.rs:209]: Nice catch

File: cli/npm/managed/cache/value_creator.rs [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+
+use deno_core::error::AnyError;
+use deno_core::futures::future::BoxFuture;
+use deno_core::futures::future::LocalBoxFuture;
+use deno_core::futures::future::Shared;
+use deno_core::futures::FutureExt;
+use deno_core::parking_lot::Mutex;
+use tokio::task::JoinError;
+
+// todo(dsherret): unit test this
```
Comment (bartlomieju) [inline cli/npm/managed/cache/value_creator.rs:13]: Yeah, unit tests here would be very useful

File: cli/npm/managed/mod.rs [hunk 9]
```diff
@@ -387,11 +388,6 @@ impl ManagedCliNpmResolver {
     self.resolution.add_package_reqs(packages).await?;
     self.fs_resolver.cache_packages().await?;
 
-    // If there's a lock file, update it with all discovered npm packages
-    if let Some(lockfile) = &self.maybe_lockfile {
-      self.lock(&mut lockfile.lock());
-    }
```
Comment (bartlomieju) [inline cli/npm/managed/mod.rs:393]: Where is lockfile handling done now?

File: cli/resolver.rs [hunk 10]
```diff
@@ -837,19 +861,50 @@ impl NpmResolver for CliGraphResolver {
...
     match &self.npm_resolver {
-      Some(npm_resolver) => match npm_resolver.as_inner() {
-        InnerCliNpmResolverRef::Managed(npm_resolver) => {
-          npm_resolver.resolve_npm_for_deno_graph(package_req)
+      Some(npm_resolver) => {
+        match npm_resolver.as_inner() {
+          InnerCliNpmResolverRef::Managed(npm_resolver) => {
+            let reqs_with_package_infos = {
+              let memory_cache = self.memory_cache.borrow();
+              package_reqs
```
Comment (bartlomieju) [inline cli/resolver.rs:874]: Nit, this becomes really hard to parse, consider doing `let Some(npm_resolver) = ... else {}` and returning early where possible. Not a blocker to landing though

File: runtime/shared.rs (hunk 4 above)
Comment (bartlomieju) [reply runtime/shared.rs:115]: This seems fine, because it's only used during build time and we want to use the fastest available "string type" for the snapshotted sources.

**Review by bartlomieju — APPROVED** [review]

File: cli/npm/managed/mod.rs (hunk 9 above)
Comment (dsherret) [reply cli/npm/managed/mod.rs:393]: This was done twice by accident. It's now updated only within the update queue of cli/npm/managed/resolution.rs (`populate_lockfile_from_snapshot`).

File: cli/module_loader.rs [hunk 11]
```diff
@@ -366,13 +366,15 @@ impl<TGraphContainer: ModuleGraphContainer>
     } else {
       self.load_prepared_module(specifier, maybe_referrer).await?
     };
+    // todo(dsherret): we should store the source maps non-inline
+    // so that we don't don't have to potentially allocate another
+    // large string when stripping the source map here
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/module_loader.rs:371]: I just looked into this a bit more right now while creating the issue and I initially thought we were doing this on an `Arc`, but actually we're doing this on an owned value, which doesn't seem like a big deal.
